### PR TITLE
Block Bindings: Change placeholder when attribute is bound

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
Adapt the placeholder and the `aria-label` attribute of the rich text component when the relevant attribute is connected to a block bindings source. The default texts don't make sense in this context.

| Before | After |
|----------|----------|
| Screenshot 2024-08-29 at 14 34 50 | Screenshot 2024-08-29 at 14 33 44 |

Same happens with the `aria-label` text, which is also changed in this PR.

Additionally, I took the opportunity to do a small refactor on `disableBoundBlock`, which is used to lock editing when appropriate, to rely on the rich-text identifier instead of iterating through all the attributes.

## Why?
It is confusing to get a prompt to insert a block when user is actually editing the binding source value.

## How?
Return an alternative placeholder when the attribute used as identifier in the RichText is connected to any source. And use this placeholder when it exists.

## Testing Instructions
1. Register a new custom field with an empty string as default value.
```
	register_meta(
		'post',
		'empty_field',
		array(
			'show_in_rest' => true,
			'single'       => true,
			'type'         => 'string',
			'default'      => '',
		)
	);
```
2. Go to a page and insert a paragraph bound to that custom field.
```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"empty_field"}}}}} -->
<p></p>
<!-- /wp:paragraph -->
```
3. Check that the placeholder shows the new text related to bindings.
4. Inspect the element and check the `aria-label` has also changed.
5. Add an empty paragraph (not bound) and check it keeps working as expected.

### CODE REVIEW

File: packages/block-editor/src/components/rich-text/index.js
Comment (SantosGuillamot) [inline packages/block-editor/src/components/rich-text/index.js]: It'd be great to get feedback on the desired text.

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [inline packages/block-editor/src/components/rich-text/index.js]:
```suggestion
			const _bindingsPlaceholder =
			// Translators: %s is the bindings source or bindings key placeholder.
			 sprintf( '%s value is empty', ( relevantBinding?.args?.key || blockBindingsSource?.label )  );
```

Comment (cbravobernal) [reply packages/block-editor/src/components/rich-text/index.js]: This should be translatable.

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [reply packages/block-editor/src/components/rich-text/index.js]: I think is fine, maybe we can remove `value`.

**Review by cbravobernal — CHANGES_REQUESTED** [review]

File: packages/block-editor/src/components/rich-text/index.js
Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: I thought about removing it, but when using the source label felt a bit weird:

* "my_custom_field is empty" sounds fine.
* "Custom Source is empty" doesn't sound as good to me.

Having said that, I don't have a strong opinion on this.

File: packages/block-editor/src/components/rich-text/index.js
Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: You're totally right. I just commited your suggestion.

File: packages/block-editor/src/components/rich-text/index.js
Comment (gziolo) [inline packages/block-editor/src/components/rich-text/index.js]: It should be a prompt for action, example:

```suggestion
				__( 'Type something to provide value for %s' ),
```

It should be rather more concise.

Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: There are two aspects to consider here:

* To simplify the workflows and the implementation, my idea was to reuse the same "empty" message when the user can edit and can not edit the meta field. For example, we are showing this message when you bind a paragraph to an empty field in a query loop, where editing is disabled.
* This message appears in every rich text, including shorter blocks like "Button". With that in mind, I would like to keep it as short as possible.

Do you think that showing " X value is empty" is not enough for prompting the user to write? Do you think that we should show different messages when the user can edit or can not?

Comment (gziolo) [reply packages/block-editor/src/components/rich-text/index.js]: Yes, these are two different use cases. In particular, these placeholder prompts calling for action were designed with accessibility in mind. It's best to check it with assistive technology to check whether it makes sense.

Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: Okay, I have just pushed a commit to differentiate between these use cases. It is now a matter of deciding the text we want to show for each case. What about this:

Use "Add {key} value" when the user can edit:

Screenshot 2024-08-30 at 15 05 52

This follows what we are doing for post title or buttons for example. It is simple, but not sure if it is clear enough.

Screenshot 2024-08-30 at 15 05 02

Use just the key (or label) when the editing is locked.

Screenshot 2024-08-30 at 15 06 15

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [reply packages/block-editor/src/components/rich-text/index.js]: Me neither 🤷‍♂️

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [reply packages/block-editor/src/components/rich-text/index.js]: Maybe just "Add empty_field"? Similar to add title, I think value is still not adding "value" 😆

Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: That could work for me 😄

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [inline packages/block-editor/src/components/rich-text/index.js:409]: is `placeholder` already defaulted on line 337?

File: packages/block-editor/src/components/rich-text/index.js
Comment (cbravobernal) [inline packages/block-editor/src/components/rich-text/index.js]: I don't really like the name, but cannot find any better. Why is relevant?

Comment (artemiomorales) [reply packages/block-editor/src/components/rich-text/index.js]: I find `relevantBinding` to be unclear.

Maybe `relatedBinding` could work? Another option is `associatedBinding`.
If neither of those, honestly for me just `binding` would work fine and might even be the clearest. Another possibility is `richTextBinding` (my least favorite, but could work). 

Other ideas: `connectedBinding`, `targetBinding`, `linkedBinding`, though I like these options less and they also seem unclear to me.

File: packages/block-editor/src/components/rich-text/index.js
Comment (artemiomorales) [reply packages/block-editor/src/components/rich-text/index.js:409]: > is placeholder already defaulted on line 337?

+1 — I think having just `aria-label={ placeholder || props[ 'aria-label' ] }` should be good.

**Review by artemiomorales — COMMENTED** [review]
Looking good! I tested and the placeholder shows as expected in the post editor and site editor. As far as I can tell, this should be good to go after addressing the points below.

File: packages/block-editor/src/components/rich-text/index.js
Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js:409]: The problem is the priority. `props[ 'aria-label' ]` is supposed to override the default placeholder value if it exist for the `aria-label` attribute. And I believe we want to override the `aria-label` prop when a binding exists.

File: packages/block-editor/src/components/rich-text/index.js
Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js]: I can use `relatedBinding` 👍

**Comment by SantosGuillamot** [comment]
I believe I've addressed the feedback you shared. This is how it is looking at this moment:

**Empty value when user CAN edit**

Screenshot 2024-09-02 at 12 21 20


**Empty value when user CAN NOT edit**

Screenshot 2024-09-02 at 12 21 59

I believe we can explore adding a label to custom fields instead of using the key in future pull requests.

Is this good enough? Something else I should change?

**Review by cbravobernal — APPROVED** [review]
LGTM
We can merge it and wait for more feedback from the community.

File: packages/block-editor/src/components/rich-text/index.js
Comment (gziolo) [reply packages/block-editor/src/components/rich-text/index.js:409]: @zaguiini, regarding your comment from https://github.com/WordPress/gutenberg/pull/65089#pullrequestreview-2283038208.

> Another issue, though related, is that the Block Bindings API does not respect an existing placeholder attribute. From the back-end, I'm setting a bespoke placeholder:

I think this is what changed the behavior, as placeholder became here a fallback. Now, the question is whether there is always `placeholder` coming from the server. If not, then we should be fine reverting to the old behavior and including the fallback to `bindingsPlaceholder` when the RichText is connected to a source.

**Pull request merged**
